pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java

```diff
@@ -60,29 +63,59 @@ public PartialUpsertHandler(Schema schema, Map<String, UpsertConfig.Strategy> pa
    * For example, overwrite merger will only override the prev value if the new value is not null.
    * Null values will override existing values if not configured. They can be ignored by using ignoreMerger.
    *
-   * @param previousRecord the last derived full record during ingestion.
+   * @param indexSegment the segment of the last derived full record during ingestion.
+   * @param docId the docId of the last derived full record during ingestion in the segment.
    * @param newRecord the new consumed record.
-   * @return a new row after merge
    */
-  public GenericRow merge(GenericRow previousRecord, GenericRow newRecord) {
-    for (String column : previousRecord.getFieldToValueMap().keySet()) {
+  public void merge(IndexSegment indexSegment, int docId, GenericRow newRecord) {
+    for (String column: indexSegment.getColumnNames()) {
       if (!_primaryKeyColumns.contains(column)) {
-        if (!previousRecord.isNullValue(column)) {
+        PartialUpsertMerger merger = _column2Mergers.getOrDefault(column, _defaultPartialUpsertMerger);
+        // Non-overwrite mergers
+        // (1) If the value of the previous is null value, skip merging and use the new value
+        // (2) Else If the value of new value is null, use the previous value (even for comparison columns).
+        // (3) Else If the column is not a comparison column, we applied the merged value to it.
+        if (!(merger instanceof OverwriteMerger)) {
+          PinotSegmentColumnReader pinotSegmentColumnReader = new PinotSegmentColumnReader(indexSegment, column);
```

Comment:
We can do try-with-resource here

Reply:
thanks for suggestion, i updated the PR with try-with-resource